src/charm.py

```diff
@@ -642,580 +92,247 @@ def k8s_auth(self) -> None:
-        """Use the v1 k8s API.
-
-        Returns:
-            The core v1 API.
-        """
-        self.k8s_auth()
-        return kubernetes.client.CoreV1Api()
-
-    def _networking_v1_api(self) -> kubernetes.client.NetworkingV1Api:
-        """Use the v1 beta1 networking API.
+    @property
+    def _endpoints_controller(self) -> EndpointsController:
+        self._k8s_auth()
+        api = kubernetes.client.CoreV1Api()
+        return EndpointsController(client=api, label=self.app.name)
 
-        Returns:
-            The networking v1 API.
-        """
-        self.k8s_auth()
-        return kubernetes.client.NetworkingV1Api()
+    @property
+    def _endpoint_slice_controller(self) -> EndpointSliceController:
+        self._k8s_auth()
+        api = kubernetes.client.DiscoveryV1Api()
+        return EndpointSliceController(client=api, label=self.app.name)
 
-    def _report_service_ips(self) -> List[str]:
-        """Report on service IP(s) and return a list of them.
+    @property
+    def _service_controller(self) -> ServiceController:
+        self._k8s_auth()
+        api = kubernetes.client.CoreV1Api()
+        return ServiceController(client=api, label=self.app.name)
 
-        Returns:
-            A list of service IPs.
-        """
-        api = self._core_v1_api()
-        services = api.list_namespaced_service(  # type: ignore[attr-defined]
-            namespace=self._namespace
-        )
-        all_k8s_service_names = [rel._k8s_service_name for rel in self._all_config_or_relations]
-        return [
-            x.spec.cluster_ip for x in services.items if x.metadata.name in all_k8s_service_names
-        ]
+    @property
+    def _ingress_controller(self) -> IngressController:
+        self._k8s_auth()
+        api = kubernetes.client.NetworkingV1Api()
+        return IngressController(client=api, label=self.app.name)
 
-    def _report_ingress_ips(self) -> List[str]:
+    def _report_ingress_ips(self, ingress: kubernetes.client.V1Ingress) -> List[str]:
         """Report on ingress IP(s) and return a list of them.
 
+        Args:
+            ingress: the target ingress.
+
         Returns:
             A list of Ingress IPs.
         """
-        api = self._networking_v1_api()
+        controller = self._ingress_controller
         # Wait up to `interval * count` seconds for ingress IPs.
-        count, interval = _report_interval_count(), 1
+        count, interval = 100, 1
+        ips = []
         for _ in range(count):
-            ingresses = api.list_namespaced_ingress(  # type: ignore[attr-defined]
-                namespace=self._namespace
+            ingresses = controller.list_resource(
+                namespace=ingress.metadata.namespace, label_selector=self._label_selector
             )
             try:
-                ips = [x.status.load_balancer.ingress[0].ip for x in ingresses.items]
+                ips = [x.status.load_balancer.ingress[0].ip for x in ingresses]
```

Comment:
From what I understand, it might be preferred to return a tuple rather than lists to keep it immutable.

Reply:
I was wondering if ips variable is needed, as in the ips can be returned right away if there are no `TypeError`s raised.